### PR TITLE
Include data streams in search shards requests

### PR DESCRIPTION
**Pull request by dnhatn**
When security is enabled, CCS requests for a remote wildcard pointing data streams are not resolved correctly. This issue happens because search_shards requests do not override the `includeDataStreams` method, causing the RBAC engine to exclude data streams while resolving the index pattern.

Relates #95894
Relates #94534

### CODE REVIEW

File: x-pack/qa/multi-cluster-search-security/legacy-with-basic-license/src/test/resources/rest-api-spec/test/fulfilling_cluster/10_basic.yml [hunk 1]
```diff
@@ -127,6 +127,8 @@ setup:
               properties:
                 '@timestamp':
                   type: date
+            settings:
+              'index.number_of_shards': 1
```
Comment (javanna) [inline x-pack/qa/multi-cluster-search-security/legacy-with-basic-license/src/test/resources/rest-api-spec/test/fulfilling_cluster/10_basic.yml:131]: why change number of shards?

**Review by javanna — APPROVED** [review]
LGTM, would it be difficult to add a unit test as well that catches how data streams are resolved for SearchShardsRequest? I am kinda nervous that we rely on yaml tests to catch that.

File: x-pack/qa/multi-cluster-search-security/legacy-with-basic-license/src/test/resources/rest-api-spec/test/fulfilling_cluster/10_basic.yml [hunk 2]
```diff
@@ -141,15 +143,36 @@ setup:
...
 
+  - do:
+      bulk:
+        refresh: true
+        index: simple-data-stream1
+        body:
+          - '{"create":{ }}'
+          - '{"@timestamp": "2023-05-31", "value": 1}'
+          - '{"create":{ }}'
+          - '{"@timestamp": "2023-07-31", "value": 2}'
```
Comment (quux00) [inline x-pack/qa/multi-cluster-search-security/legacy-with-basic-license/src/test/resources/rest-api-spec/test/fulfilling_cluster/10_basic.yml:162]: Does including `@timestamp` automatically make it a data streams index?

File: x-pack/qa/multi-cluster-search-security/legacy-with-basic-license/src/test/resources/rest-api-spec/test/querying_cluster/10_basic.yml [hunk 3]
```diff
@@ -352,3 +352,40 @@ teardown:
   - match: { acknowledged:   true }
 
 
+---
+"Search against remote data streams":
+  - do:
+      headers: { Authorization: "Basic am9lOnMza3JpdC1wYXNzd29yZA==" }
+      search:
+        index: '*:simple-data-stream*'
```
Comment (quux00) [inline x-pack/qa/multi-cluster-search-security/legacy-with-basic-license/src/test/resources/rest-api-spec/test/querying_cluster/10_basic.yml:360]: Should you add `ccs_minimize_roundtrips: true` here to make it clear what you are testing (compared to the next test)?

**Comment by quux00** [comment]
Should the commit message also mention that it addresses https://github.com/elastic/elasticsearch-benchmarks/issues/1554 ?

**Comment by dnhatn** [comment]
> would it be difficult to add a unit test as well that catches how data streams are resolved for SearchShardsRequest

++, I added a unit test in https://github.com/elastic/elasticsearch/pull/96469/commits/0c99c83df6dfd79ccbc0c83e673472428f52f963. This requires security.

File: x-pack/qa/multi-cluster-search-security/legacy-with-basic-license/src/test/resources/rest-api-spec/test/querying_cluster/10_basic.yml (hunk 3 above)
Comment (dnhatn) [reply x-pack/qa/multi-cluster-search-security/legacy-with-basic-license/src/test/resources/rest-api-spec/test/querying_cluster/10_basic.yml:360]: I think we prefer the implicit default; but I pushed https://github.com/elastic/elasticsearch/pull/96469/commits/93fcddb3878412c10aac1a19449a1b2095c3028e.

File: x-pack/qa/multi-cluster-search-security/legacy-with-basic-license/src/test/resources/rest-api-spec/test/querying_cluster/10_basic.yml [hunk 4]
```diff
@@ -352,3 +352,40 @@ teardown:
...
+        body:
+          query:
+            range:
+              "@timestamp":
+                gte: "2023-01-01"
+  - match: {_clusters.total: 1}
+  - match: {_clusters.successful: 1}
+  - match: {_clusters.skipped: 0}
+  - match: { _shards.total: 3 }
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply x-pack/qa/multi-cluster-search-security/legacy-with-basic-license/src/test/resources/rest-api-spec/test/querying_cluster/10_basic.yml:370]: We can't have the same assertion; but I added `skipped==0` and a comment explaining it in https://github.com/elastic/elasticsearch/pull/96469/commits/93fcddb3878412c10aac1a19449a1b2095c3028e.

File: x-pack/qa/multi-cluster-search-security/legacy-with-basic-license/src/test/resources/rest-api-spec/test/fulfilling_cluster/10_basic.yml (hunk 1 above)
Comment (dnhatn) [reply x-pack/qa/multi-cluster-search-security/legacy-with-basic-license/src/test/resources/rest-api-spec/test/fulfilling_cluster/10_basic.yml:131]: I need to assert the number of shards that search requests with data streams executed against. The default the number of shards is 1; but we might random it in YAML tests.

**Comment by dnhatn** [comment]
@javanna @quux00 Thanks for reviews.

**Pull request merged**